tasks/main.yml

```diff
@@ -110,6 +101,10 @@
         - logging_debug | d(false)
 
     - name: Include Rsyslog role
+      vars:
+        __rsyslog_enabled: "{{ logging_enabled | d(true) }}"
+        __rsyslog_purge_original_conf: "{{ logging_purge_confs | d(false) }}"
+        __rsyslog_system_log_dir: "{{ logging_system_log_dir | d('/var/log') }}"
```

Comment:
Not sure if we've had this discussion before, so forgive me and remind me if we have, or if this is different since rsyslog is a sub-role of logging, but . . .

. . . you don't need the `| d()` filter for these since they are defined in `logging/defaults/main.yml` - @pcahyna please correct me if I'm wrong, but you can do this:
```yaml

      vars:
        __rsyslog_enabled: "{{ logging_enabled }}"
        __rsyslog_purge_original_conf: "{{ logging_purge_confs }}"
        __rsyslog_system_log_dir: "{{ logging_system_log_dir }}"
```
again, if I'm forgetting some discussion we already had about this, let me know

Reply:
@richm, @pcahyna, sorry, give me one sec.  I thought I should be able to remove the default filters safely, but I'm observing some test failures and need to debug it...
I set hold label.

Reply:
@richm, @pcahyna, I pushed the [3rd patch](https://github.com/linux-system-roles/logging/pull/99/commits/d62a7dff24cd1810ad1fd9e886079d6426147d29
) following the suggestions from @richm.

Not just removing the default filter from the vars, I had to move the default setting `logging_system_log_dir: /var/log` from the top level defaults/main.yml to the rsyslog level roles/rsyslog/defaults/main.yml.  It was needed due to the logging specific role structure (please see also https://github.com/linux-system-roles/logging/blob/master/design_docs/logging_subsystem.md).

The top level tasks/main.yml includes the rsyslog role (https://github.com/nhosoi/logging/blob/rsyslog_system_log_dir/tasks/main.yml#L108-L109), then in the rsyslog role, eventually each sub-role (such as https://github.com/nhosoi/logging/tree/rsyslog_system_log_dir/roles/rsyslog/roles/input_roles/basics and https://github.com/nhosoi/logging/tree/rsyslog_system_log_dir/roles/rsyslog/roles/output_roles/files) includes https://github.com/nhosoi/logging/blob/rsyslog_system_log_dir/roles/rsyslog/tasks/deploy.yml like this (https://github.com/nhosoi/logging/blob/rsyslog_system_log_dir/roles/rsyslog/roles/output_roles/files/tasks/main.yml#L7-L9).  Usually, if the default value is assigned to a variable in the top level defaults/main.yml, it should be available in any place underneath, I suppose. But it's not after 2 step of including's.  Unless I move `logging_system_log_dir` to roles/rsyslog/defaults/main.yml, it fails with `logging_system_log_dir` undefined.

If the move is not acceptable, there are 2 alternatives.
1) moving back to set_fact.
2) giving up to have rsyslog internal vars, and exposing rsyslog vars as external and using them in the rsyslog roles.

Do you think what'd be the best?  Thanks!

Reply:
Reminder to myself: I have to do the same thing on __rsyslog_purge_original_conf, too.  (maybe in another PR)

Reply:
So that means `logging_system_log_dir` set in the top level `defaults/main.yml` is undefined by the time the rsyslog role is called?  So at this point: 
```yaml
      vars:
        __rsyslog_enabled: "{{ logging_enabled }}"
        __rsyslog_purge_original_conf: "{{ logging_purge_confs }}"
        __rsyslog_system_log_dir: "{{ logging_system_log_dir }}"
```
this means that `logging_system_log_dir` is always undefined?  So what is the point of using it at all?  Are you saying that if you use `set_fact` at the top level to set `logging_system_log_dir`, it will be defined at the point above?  @greg-hellings have you ever seen anything like this before?  @maxamillion is this some sort of strange bug in ansible?

Reply:
`set_fact` has exceptionally high priority for values (I think only command-line defined parameters override it, if even that?) and are set globally throughout the entire rest of a playbook run. So it would make sense that using `set_fact` works. But also, because of that, using `set_fact` is, generally, a terrible idea. Especially for something that's just a parameter to a role.

I can't comment on what's "usual" when it comes to embedded roles. We steadfastly refuse to use them anywhere in OASIS, and I haven't yet encountered a reason to use them that wasn't covered just as well by promoting the subrole to a top level role. In the old days, before collections, that could create dependency problems. But now that we're in the Collections era, why not convert this to a collection, have this role and all its children as peer roles, and include them from within the same collection? If there really is a reason that these roles don't deserve to be first-class citizens, then why are they roles at all and not just task import/includes?

Reply:
> So that means `logging_system_log_dir` set in the top level `defaults/main.yml` is undefined by the time the rsyslog role is called? So at this point:
<< snip >>
> this means that `logging_system_log_dir` is always undefined? So what is the point of using it at all?

If you set logging_system_log_dir in your inventory file, the value is respected.
https://github.com/nhosoi/logging/blob/rsyslog_system_log_dir/tests/tests_default_files_log_dir.yml#L6

Reply:
Thank you,  @greg-hellings.
> But now that we're in the Collections era, why not convert this to a collection, have this role and all its children as peer roles, and include them from within the same collection? If there really is a reason that these roles don't deserve to be first-class citizens, then why are they roles at all and not just task import/includes?

I've never used collection yet...  Could you please give me pointer(s) to learn how to do the conversion?  Thanks!

Reply:
We are currently planning about how and when to convert all of the linux-system-roles to collections.  So don't worry about this for now for logging (although if you want to learn about ansible collections, please do).

As for the current problem, it seems that the only way to have `logging_system_log_dir` in scope when the rsyslog sub-role is called is to use `set_fact`, so please do that, and create another issue to revisit this.

Reply:
I pushed this patch to use set_fact.  Could you review it one more time, @richm, @pcahyna?
Thanks!!
https://github.com/linux-system-roles/logging/pull/99/commits/b440c7d9504744fa1cd057fafe75b4f32c4c6742

Reply:
> As for the current problem, it seems that the only way to have `logging_system_log_dir` in scope when the rsyslog sub-role is called is to use `set_fact`, so please do that, and create another issue to revisit this.

Would it work to keep the variable in defaults and not moving it to vars?

Reply:
> If the move is not acceptable, there are 2 alternatives.
> 
>     1. moving back to set_fact.
> 
>     2. giving up to have rsyslog internal vars, and exposing rsyslog vars as external and using them in the rsyslog roles.

I am interested what option 2 means. Where would be rsyslog vars exposed? In which defaults.yml file? Of the `rsyslog` role?

Reply:
@nhosoi @greg-hellings I found that the problem was not due to embedded roles (lexically nested), but in dynamic nesting in the role call chain, specifically, when a role calls indirectly itself (indirect recursion). So moving from embedded roles to a flat structure will not help. One will have to stop letting `rsyslog` call `basics` and `basics` call back `rsyslog`. This is shown in a minimal reproducer here: https://github.com/pcahyna/nested_roles_demo

Reply:
Dropping this here to make sure it's been consulted, following the thread it seems the like the var precedence was called into question. https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#ansible-variable-precedence

@richm I'd need an example playbook and the corresponding roles that are showing the issue but it's entirely possible and I'd be interested in taking a look. We can follow-up off-thread if you like.

Reply:
@maxamillion example playbook and corresponding roles is here: https://github.com/pcahyna/nested_roles_demo/blob/master/logging_flattened/testplaybook.yml . See https://github.com/pcahyna/nested_roles_demo/blob/master/README for a brief explanation.

Reply:
@nhosoi I suppose the issue would be fixed by #108 and `set_fact` could be eliminated?

Reply:
> @nhosoi I suppose the issue would be fixed by #108 and `set_fact` could be eliminated?

@pcahyna, let me test it...

Reply:
> @nhosoi I suppose the issue would be fixed by #108 and `set_fact` could be eliminated?

@pcahyna, yes, it is fixed by https://github.com/linux-system-roles/logging/pull/108.  Let me apply the patch to the master branch.  Thanks.